0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+  
+  This balance can be refilled by calling `withdraw` function.
+* `burnt_balance` -- the balance that the contract already used for:
+  * Wasm exeuction fees;
+  * Fees for calling host functions, including fees for creating promises, **excluding balance attached to these promises.**
+  
+  This balance cannot be recovered, if the contract execution fails.
+  
+When contract execution fails the following happens to the balances:
+* First we compute `deficit = all_withdrawals - all_deposits`;
+* If `deficit > 0` (we have withdrawn from account more than we deposited to it):
+  * If `burnt_balance <= attached_balance` then we deposit `deficit` back to the `account_balance` and issue
+    a refund of `attached_balance - burnt_balance` if this value is not zero.
+  * If `attached_balance < burnt_balance <= attached_balance + deficit` then we deposit
+  `attached_balance + deficit - burnt_balance` back to the `account_balance` and do not issue a refund;
+* If `deficit <= 0` (we have deposited to the account more than what we have withdrawn) we withdraw `abs(deficit)`
+  from the `account_balance` and issue a refund of `attached_balance - burnt_balance`.
+  
+Note, for both cases above, `burnt_balanace` is never greater than `attached_balance + deficit` at any point in the program. This
+might require a formal proof later.
+
+---
+The following spec is the same for all functions:
+```
+account_balance(balance_ptr: u64)
+attached_balance(balance_ptr: u64)
+remaining_balance(balance_ptr: u64)
+burnt_balance(balance_ptr: u64)
+```
+ -- writes the value into the `u128` variable pointed by `balance_ptr`.
+
+###### Panics
+* If `balance_ptr + 16` points outside the memory of the guest with `MemoryAccessViolation`;
+
+###### Current bugs
+* Use a different name;
+* Unclear how they behave with the callbacks.
+
+---
+```rust
+deposit(min_amount_ptr: u64, max_amount_ptr: u64, result_ptr: u64)
+```
+Moves balance from `attached_balance` to `account_balance`.
+If `min_amount <= attached_balance` moves `min(max_amount, max(min_amount, attached_balance))`. The moved amount is
+saved into `u128` variable pointed by `result_ptr`. `min_amount_ptr` and `max_amount_ptr` point to `u128` variables.
+
+###### Panics
+* If `min_amount_ptr + 16` or `max_amount_ptr + 16` or `result_ptr + 16` point outside the memory of the guest with `MemoryAccessViolation`;
+
+---
+
+```rust
+withdraw(min_amount_ptr: u64, max_amount_ptr: u64, result_ptr: u64)
+```
+Moves balance from `account_balance` to `attached_balance`.
+If `min_amount <= attached_balance` moves `min(max_amount, max(min_amount, account_balance))`. The moved amount is
+saved into `u128` variable pointed by `result_ptr`. `min_amount_ptr` and `max_amount_ptr` point to `u128` variables.
+
+###### Panics
+* If `min_amount_ptr + 16` or `max_amount_ptr + 16` or `result_ptr + 16` point outside the memory of the guest with `MemoryAccessViolation`;
+
+## Math
+
+```rust
+random_buf(buf_len: u64, buf_ptr: u64)
```

Comment:
It makes sense to switch to writing into a new register instead to be consistent.
`random_buf(len: u64, register_id: u64)`
The `register_id` is where to write

Reply:
Done